internal/event/stream.go

```diff
@@ -66,88 +80,63 @@ func (s *Stream) start() context.CancelFunc {
 }
 
 // process manages an Stream's lifecycle until canceled by the provided
-// context.  Incoming events are appended to a queue which is then relayed to
+// context. Incoming events are appended to a queue which is then relayed to
 // the a listening client. New events pushed onto the queue will not block due
 // to client behavior.
 func (s *Stream) process(ctx context.Context) {
+	// Close the output channel so that clients know this stream is finished.
+	// We don't close s.in to avoid creating a race with the stream's Push()
+	// function.
+	defer close(s.out)
+
 	for {
-		// informs send() to stop trying
-		nctx, next := context.WithCancel(ctx)
-		defer next()
+		// This loop relies on the fact that a send to a nil channel will block
+		// forever. If we have no entries in the queue, we set the send channel
+		// to nil, so the clause that attempts to send an event to the client
+		// will never complete. Clients will never receive an emptyEvent.
+		sendCh := chan Event(nil)
+		next := Event(emptyEvent{})
```

Comment:
i now see how this  is clever, but I couldn't understand the purpose until seeing the simplified `<s.send(...)` case below. could just be me. `empty` might be a better name? I was confused, thinking we were somehow signaling to the client that the queue was empty upon receipt of an `emptyEvent` (which i realize makes no sense).

Reply:
Ugh. Clever is a 4-letter word. I'm restructuring this slightly and updating the comments to hopefully make it clearer what's going on here.